packages/core/streams/openai-stream.ts

```diff
@@ -10,6 +10,101 @@ function parseOpenAIStream(): (data: string) => string | void {
 
+    /*
+       If the response is a function call, the first streaming chunk from OpenAI returns the name of the function like so
+
+          {
+            ...
+            "choices": [{
+              "index": 0,
+              "delta": {
+                "role": "assistant",
+                "content": null,
+                "function_call": {
+                  "name": "get_current_weather",
+                  "arguments": ""
+                }
+              },
+              "finish_reason": null
+            }]
+          }
+
+       Then, it begins streaming the arguments for the function call.
+       The second chunk looks like:
+
+          {
+            ...
+            "choices": [{
+              "index": 0,
+              "delta": {
+                "function_call": {
+                  "arguments": "{\n"
+                }
+              },
+              "finish_reason": null
+            }]
+          }
+
+        Third chunk:
+
+          {
+            ...
+            "choices": [{
+              "index": 0,
+              "delta": {
+                "function_call": {
+                  "arguments": "\"location"
+                }
+              },
+              "finish_reason": null
+            }]
+          }
+
+        ...
+
+        Finally, the last chunk has a `finish_reason` of `function_call`:
+
+          {
+            ...
+            "choices": [{
+              "index": 0,
+              "delta": {},
+              "finish_reason": "function_call"
+            }]
+          }
+
+
+        With the implementation below, the client will end up getting a
+        response like the one below streamed to them whenever a function call
+        response is returned:
+
+          {
+            "function_call": {
+              "name": "get_current_weather",
+              "arguments": "{\"location\": \"San Francisco, CA\", \"format\": \"celsius\"}
+            }
+          }
+     */
+    if (json.choices[0]?.delta?.function_call?.name) {
+      return `{"function_call": {"name": "${json.choices[0]?.delta?.function_call.name}", "arguments": "`
+    } else if (json.choices[0]?.delta?.function_call?.arguments) {
+      console.log(`json.choices[0]?.delta?.function_call.arguments: ${json.choices[0]?.delta?.function_call.arguments}`)
+      const argumentChunk: string = json.choices[0].delta.function_call.arguments;
+
+      let escapedPartialJson = argumentChunk
+        .replace(/\\/g, '\\\\') // Replace backslashes first to prevent double escaping
+        .replace(/\//g, '\\/')  // Escape slashes
+        .replace(/"/g, '\\"')   // Escape double quotes
+        .replace(/\n/g, '\\n')  // Escape new lines
+        .replace(/\r/g, '\\r')  // Escape carriage returns
+        .replace(/\t/g, '\\t')  // Escape tabs
+        .replace(/\f/g, '\\f'); // Escape form feeds
+
+      return `${escapedPartialJson}`
+    } else if (json.choices[0]?.finish_reason === 'function_call') {
+      return '"}}'
+    }
```

Comment:
Should we move this logic to another function?  🤔

Reply:
Would second this, maybe with a test too.

Reply:
@Danm72 I'm not sure if those replacement statements should be here. Additionally, it seems to be the client's responsibility, since the argument chucks may be different depending on the type of function we pass to Open AI. Thoughts?

Reply:
@lautapercuspain  I agree that it's a little weird to have these replace statements, and there may be a better alternative. The issue is that without the character escaping, parsing the function call object with JSON.parse does not work.. I'll try to think of a better way to do this, but would appreciate suggestions as well.

Reply:
@Zakinator123. I'm currently debugging your PR 👍🏻.   
So far the `get_current_weather` and `get_current_time` works in a reliable and predictable way.  
The issue with JSON.parse raises when it comes to code execution, like your example with `eval` which would be really cool to support out of the box on the other hand. Maybe the simplest solution here is to instruct the agent (as you are doing in your prompt to escape special characters, when it comes to code functionality.
I'll comment any findings on my side.